Add docs for submitting code contributions

UniversalSuperBox opened the pull request:
We've recently had trouble with people doing QA asking code contributors
how they should test a PR, or what issues are fixed by a PR. We've also
had a lot of questions about what makes a good contribution to Ubuntu
Touch. Hopefully this information can prevent and clear up these
questions in the future.

All of the reviewers listed must approve before this is merged.

mardy commented on contribute/code.rst:
```diff
@@ -0,0 +1,158 @@
+Code
+====
+
+We're serious about code quality and stability. We have to be, Ubuntu Touch is used by many users on many different devices. Our quality requirements be intimidating for new contributors as we often ask for many rounds of changes during reviews. We understand that you've chosen to spend your time developing changes to Ubuntu Touch as a volunteer, and we thank you for that. We will respect you and ...
```
"be intimidating" -> "might be intimidating", I guess

UniversalSuperBox replied:
Arr, you don't like our review style matey?

doniks replied:
should that be a full stop after "We have to be." or maybe even an exclamation mark

mardy commented on contribute/code.rst:
```diff
@@ -0,0 +1,158 @@
+
+* In ``ubports/core/docs!1``::
+
+    https://gitlab.com/ubports/core/code!2 and https://gitlab.com/ubports/core/infrastructure!3 depend on this MR.
```
It would be better to split these lines, or to use a different formatting style so that they are automatically wrapped in the rendered view. Currently one has to scroll in order to see the end of the line.

mardy reviewed (approved):
LGTM, just a tiny issue.

BTW, I'm not sure how it's done, but I believe it might be possible to provide a commit template, so that the most important guidelines are shown as comments in the editor when someone submits a commit. Though I'm not sure if it can be done for the whole organisation, or if the commit template must be present in every repo (in that case, I'm quire sure we shouldn't do it :-D )

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,158 @@
+
+When you receive a request for changes that affects the first commit in your series, edit that commit instead of adding a new one on top. If doing this changes the information in your commit message, update that too. Make sure that the future traveler's job is as easy as possible.
+
+You will have to learn how to use Git to edit the series of commits in your change and upload your new series. You will generally edit your commit series with ``git rebase``. Then you will force-push your new series over the old one or make a new branch and open a new changeset. These are not perfect solutions, owing to the design of Git, GitHub, and GitLab... but force pushes or new changesets a ...
```
should we suggest `git rebase -i`?

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,158 @@
+
+You will have to learn how to use Git to edit the series of commits in your change and upload your new series. You will generally edit your commit series with ``git rebase``. Then you will force-push your new series over the old one or make a new branch and open a new changeset. These are not perfect solutions, owing to the design of Git, GitHub, and GitLab... but force pushes or new changesets a ...
+
+Luckily, there are graphical tools that make this job easier. Unluckily, if you don't understand what they are doing on the Git command line, you might mess up and have to start over from your old series. Even seasoned Git pros mess up this process sometimes, don't worry. Just know that there is almost always a way back to your old series.
```
sometimes, don't worry. -> sometimes. Don't worry.

doniks reviewed (approved):
lgtm. couple small (optional) items below

UniversalSuperBox commented:
I believe I have resolved the listed problems (and added new ones) by restructuring the document slightly. I've been careful to reduce wordiness and use active language. I've also added more information about how our review process will work.

mardy reviewed (dismissed):
LGTM, just click the "Ready to review" button, to remove the "Draft" label.

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,213 @@
+Reasoning
+---------
+
+Ubuntu Touch is used by many users on many different devices. Breaking functionality that these users rely on can be catastrophic. Our changeset review process helps us avoid lasting broken functionality by ensuring a few properties of the change.
```
I think the "lasting" is a leftover? And I think the "changeset" is just double wordy. How about "Our review process helps us avoid broken ..."

UniversalSuperBox replied:
Changeset is duplicated, but "lasting broken" meaning "broken for more than a day or two." I agree it's confusing, let's drop it.

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,213 @@
+* cause any unexpected behavior?
+* make it easy for someone to make further changes to the software in the future?
+
+Sometimes we make mistakes. We may accept a change which causes unexpected behavior or does not meet its own goals. In that case, the other properties can help us decide what to do next. We might revert the changeset, going back to the state before the change was made. We might also build on top of the old changeset. The decision can be made quickly and without breaking more functionality as long ...
```
I'm lost with this paragraph. What are we trying to say here? Seems it might be explaining some corner cases/exceptions. But we haven't even explained the big picture. Also sure there will always be exceptions. And exceptions to the exceptions ....  Bottom line, I don't understand why this paragraph is needed, I'd be fine with dropping it entirely

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,213 @@
+
+The rules on this page explain how you should write and document your changes to Ubuntu Touch. Following these rules makes the review process easier for us. An easier review means your changes are accepted and deployed to the many users of Ubuntu Touch more quickly.
+
+These are not absolute; we won't automatically reject your change based on whether you follow the rules. Make sure you can explain your decision to ignore one or more rules when asked, though.
```
I find it easier to understand when turned around: 

"Make sure .... though" -> "If you break a rule, make sure you can explain why."

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,213 @@
+When?
+    Knowing the events surrounding a change's authoring can be just as important as knowing the date of authorship. Git will automatically embed the date you created your commit into the commit itself. It's up to you to provide any temporal context, such as being asked to author a change during an in-person event or a difficult Agile-style sprint.
+Why?
+    This is the most important question to answer in your commit message. You should always state the reasoning behind a change: the bugs it fixes, the features it builds up to, and especially the limitations that cause it to look more complicated than a viewer might expect. Provide permanent links to issue reports or documentation if they help you make your case. The first question someone will  ...
```
maybe sort in order of importance?! smt like what, why, who, where when

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,213 @@
+We may ask for huge changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This is distinct from the previous section. Sometimes your reviewers will look at your change and realize that a much deeper issue exists in the component. Sometimes you start out developing a feature with the wrong assumptions in mind and end up with an unwieldy mess of spaghetti code. Either way: you took the wrong path to your destination. Whatever the reasons may be, your reviewer will polite ...
```
maybe the following could take out some pressure, by making it less personal: "you took the wrong path to your destination" -> "this is the wrong path to the destination"

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,213 @@
+You can tell us that we are wrong
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You might be sure that an underlying bug is not easily fixable, or that the path you've taken to your destination is the most correct one possible. You might not have enough time to do a fix correctly as asked. If you think that our rejection of your changeset is wrong for any reason, let us know. Reviewers will try to work with you to reach a solution that's best for our users.
```
related to prev comment, and simpler: "that the path ... possible" -> "that the path you've taken is the best"

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,213 @@
+Merge and continuing maintenance
+--------------------------------
+
+After we have ensured your changeset meets all of our requirements, it will become a part of Ubuntu Touch. Congratulations! You are a member of a small group of people contributing to a beautiful community. The work is not necessarily done, though.
```
mhm ... maybe "Thank you!" instead of "Congratulations!"?

doniks commented on contribute/code.rst:
```diff
@@ -0,0 +1,213 @@
+
+Sometimes your changes will break a piece of Ubuntu Touch functionality despite our best efforts. If bugs appear around the time your change is merged, you might be called upon to help investigate the source of the issue and prepare a fix. If we find that your change was the direct cause of a bug and we are unable to get in contact with you, your change could be reverted.
+
+New contributors might notice that you have recently worked on a component that they want to work on too. We would be honored if you helped teach them what you learned during this process.
```
maybe "grateful" rather than "honored". Or "It would be great"

Danfro commented on contribute/code.rst:
```diff
@@ -0,0 +1,211 @@
+Code changes
+============
+
+This page helps you get your changes to Ubuntu Touch's code merged. To help you with this, we describe the things we look for in code, commits, and changesets. We also take a look at what to expect in the code review process.
```
To help you with this -> To assist you with this

removes a double "help you" in this paragraph

Danfro commented on contribute/code.rst:
```diff
@@ -0,0 +1,211 @@
+    A GitHub Pull Request (PR) or GitLab Merge Request (MR).
+issue report
+    A GitHub or GitLab Issue.
+
```
Both gitlab and github use the word "issues". So this does not need explaining. But if we state "issue report", meaning no other issues or reports anywhere else, shouldn't we also name "commit" here too?

commit
           A code contribution that is pushed to either a Gitlab or Github repository.

Danfro commented on contribute/code.rst:
```diff
@@ -0,0 +1,211 @@
+* meet its own goals, such as adding a feature or fixing a bug?
+* cause any unexpected behavior?
+* make it easy for someone to make further changes to the software in the future?
+
```
I am not 100% sure if there are meanings of "Reasoning" I don't know. So I might be wrong with this one. But I think questions are not "giving reasons".

If kept as questions, I think it ("In particular..." up to "...in the future?") would be better placed somewhere under commit description maybe.

Or rephrase to aims since we want to give reasons:

In particular we want to make sure the change does...
* meets its own goals, such as adding a feature or fixing a bug
* does not cause any unexpected behavior
* makes it easy for someone to make further changes to the software in the future

Danfro commented on contribute/code.rst:
```diff
@@ -0,0 +1,211 @@
+Some people also add "How?" to their list of Ws, but the content of your commit rather than its message should answer that question.
+
+It is possible to provide some of this context in code comments rather than in commit messages. How you do it is up to you, but remember that comments can easily become out-of-date compared to the code that they are near. A commit message is linked only to the code you wrote.
+
```
It is possible to provide some of this context in code comments rather than in commit messages. Just bear in mind that comments can easily become out-of-date compared to the code that they are near. A commit message is linked only to the code you wrote. On the other hand some information need to be given to future developers who only look at the code but do not check the commit history.

Danfro commented on contribute/code.rst:
```diff
@@ -0,0 +1,211 @@
+    Knowing the events surrounding a change's authoring can be just as important as knowing the date of authorship. Git will automatically embed the date you created your commit into the commit itself. It's up to you to provide any temporal context, such as being asked to author a change during an in-person event or a difficult Agile-style sprint.
+
+Some people also add "How?" to their list of Ws, but the content of your commit rather than its message should answer that question.
+
```
Hmm. Do we need to stick to 5 W's? The last two seem strangs and rather irrelevant in most cases to me. The "how" I think is more helpful.
We could maybe drop the last two W's and use a header more like this:

Write commit messages that answer some crucial questions

Danfro commented on contribute/code.rst:
```diff
@@ -0,0 +1,211 @@
+Ensure that your changeset answers the following questions in its description.
+
+If you've followed the guidelines for commit messages, you can probably copy the relevant information from the commit messages into your changeset description. GitHub and GitLab will do this for you in single-commit changesets. A description that says "this is complicated to explain, please read the commit messages" is also fine. We trust you to strike the balance.
+
```
I think a note about adding screenshots would be good here, maybe:

If you did any changes regarding the user interface, provide screenshots showing the appearance  before and after.

Danfro commented on contribute/code.rst:
```diff
@@ -0,0 +1,211 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+All changes should be linked to an issue report that they resolve. It does not matter if you are adding a feature or fixing a problem. Is there an issue report filed on any repository which describes the problem you're fixing or the feature you're adding? Does the report specify which devices the issue occurs on or where the feature applies? Does it provide enough information that someone could l ...
+
```
Here I would suggest to use demands rather than questions:

All changes should be linked to the issue report that they resolve. It does not matter if you are adding a feature or fixing a problem. Add a link in your changeset description to the issue report filed on any repository which describes the problem you're fixing or the feature you're adding. Specify the devices the issue occurs on or if it applies to all devices. Provide enough information so that someone can look at it and knows how to reproduce the problem or test the feature. If those information are not given in the issue report, add them!

Danfro commented on contribute/code.rst:
```diff
@@ -0,0 +1,211 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+All changes require testing to ensure they resolve the issue report they're referencing. On which devices did you test your change? How did you do that? Do you have any questions about existing functionality that you are unable to test? Make sure you include this in your changeset description. If you figured out how to reproduce an issue that has long been a mystery, add that to the issue report  ...
+
```
Same here:

All changes require testing to ensure they resolve the issue report they're referencing. Describe on which devices you did test your changes. Explain how you did did that and which channel (stable, release candidate, devel, edge) you where using. Name any questions about existing functionality that you where unable to test.

Danfro reviewed (commented):
All in all a very comprehensive text. Well done. 

The only downside I can see is the overall length. Not many may read all of it... :thinking: On the other hand, it is the docs, not a cheatsheet.

Flohack74 reviewed (approved):
Merge whenever you like!

doniks commented:
@dobey, @peat-psuwit approve?

Flohack74 commented:
I will merge this now, after all its just a doc and not code, and we can change it any minute. @UniversalSuperBox thx so much for the effort and correct native English elegancy :)
